Scaling the Waku Protocol: A Performance Analysis with Wakurtosis (#123)
vacp2p/vacp2p.github.io: rlog/2023-08-04-wakurtosis-waku-scallability.mdx

&mdash;
layout: post
name: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'  
title: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'
date: 2023-08-08 12:00:00
authors: Daimakaimura
published: true
slug: wakurtosis-waku-scallability-simulations
categories: waku, wakurtosis
&mdash;

# Scaling the Waku Protocol: A Performance Analysis with Wakurtosis

## Introduction

[Waku](https://waku.org/) is a family of P2P protocols enabling private, metadata-resistant messaging for Web3 by providing censorship resistance, adaptability, modular design, and a shared service network. Waku is designed to enable communication between decentralized applications (dApps) in a peer-to-peer manner.  
It serves as an improvement and successor to Ethereum's Whisper protocol, offering better scalability and efficiency.

> [Waku Relay](https://rfc.vac.dev/spec/11/), on the other hand, is a specific component or sub-protocol within the broader Waku framework.  

Author: Fixed

It is responsible for relaying messages between nodes in the Waku network, effectively serving as the message dissemination mechanism.  
While Waku encompasses a variety of functionalities and improvements for decentralized messaging, Waku Relay specifically focuses on the message propagation aspect within this larger system.

Finally, [Discv5 (Discovery v5)](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md) is a peer-to-peer networking protocol designed to facilitate node discovery and routing in Ethereum-based decentralized networks. It serves as an upgrade to earlier discovery protocols and is designed to be modular and extensible, allowing it to support various types of decentralized systems beyond just Ethereum.  
The protocol enables nodes to find each other and maintain a distributed hash table for efficient routing.

> The scalability and performance of the Waku protocol are of critical importance. To explore these facets with high granularity across a wide range of scenarios, we turned to [Wakurtosis](https://github.com/vacp2p/wakurtosis), a bespoke simulation framework developed internally.  

Author: Done "Unfortunately, Wakurtosis did not fulfill all our requirements. We will delve into the specifics of these shortcomings in a retrospective article coming shortly"

### Scalability  

To overcome scalability challenges, Wakurtosis employs a multi-node approach, running several nodes within one container.
This method supports simulations with over 1,000 nodes on a single machine.  
However, this can introduce unforeseen network effects, potentially affecting some metrics. For instance, running several nodes per container can alter propagation times, as nodes grouped within the same container may exhibit different messaging behavior compared to a true node-to-node topology.
Additionally, employing a multi-node approach may result in losing node-level sampling granularity depending on the metrics infrastructure used, e.g. Cadvisor.

Nevertheless, Wakurtosis offers the flexibility to choose between this and a 1-to-1 simulation, catering to the specific needs of each test scenario. The results presented here are 1-to-1 simulations.  

## Examining the Waku Protocol

### Simulation Setup

To evaluate Waku under varied conditions, we conducted simulations across a range of network sizes, topologies, and message rates. Each simulation lasted 3 hours to reach a steady state.

> The network sizes explored included 75, 150, 300, and 600 nodes. For Non-Discv5 simulations, we used static topologies with average node degrees of K=50. In simulations with Discv5, we set the max_peers parameter to 50 to approximate similar average degrees.

Author: There is an explanation in the subsection *Clarification on Non-Discv5 Scenarios* bellow. Also, Discv5 is briefly explained in the intro.

To stress test message throughput, we simulated message rates of 1, and 10 messages per second. We initally run simulations up to a 100 msg/s but we found out the results were unreliable due to simulation hardware limitations and therefore decided not to include them in this analysis.

> We also included simulations batches with no load &mdash; i.e. 0 Msg/s. &mdash; to provide a clearer picture of Waku's baseline resource demands and inherent overhead costs stemming from core protocol operations.

Author: Not sure if I follow, can you please elaborate a bit?

In a live environment, some form of peer discovery mechanism would be necessary for the functioning of the network.
Mechanisms like 'rendezvous' would also introduce additional bandwidth costs.

Therefore, the intention behind including Non-Discv5 simulations is not to compare their performance directly with Discv5 but rather to establish a fundamental baseline against which the added complexities and scaling costs of employing a discovery mechanism like Discv5 can be better understood.  

### Simulation Results

To provide a clear visualization of bandwidth usage in all the cases we studied, we've presented the results in two plots, one for the non-discovery case and a second one for the discovery case. Both plots are divided in two subplots, one for reception (Rx) and one for transmission (Tx):

![Bandwidth efficiencies over adjusted pure payloads (baseline)](/static/img/wakurtosis_waku/Baseline_Efficiency.png)

![Bandwidth efficiencies over adjusted pure payloads (Discv5)](/static/img/wakurtosis_waku/Discv5_Efficiency.png)  

In these plots, efficiency is determined by contrasting the actual bytes transmitted (or received) due to messages with the total bandwidth consumed. 

> It's essential to highlight that for better accuracy, we consider the actual message counts that were successfully injected into the network since due to hardware limitations, we've observed that not all anticipated messages get injected, which can skew the interpretation of bandwidth usage. 

Author: Switched plots to total bandwidth usage.

&mdash;
layout: post
name: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'  
title: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'
date: 2023-10-19 12:00:00
authors: Daimakaimura
published: true
slug: wakurtosis-waku-scallability-simulations
categories: waku, wakurtosis
&mdash;

# Scaling the Waku Protocol: A Performance Analysis with Wakurtosis

## Introduction

> [Waku](https://waku.org/) is a family of P2P protocols enabling private, metadata-resistant messaging for Web3 by providing censorship resistance, adaptability, modular design, and a shared service network. 

Reviewer: ```suggestion
[Waku](https://waku.org/) is a family of P2P protocols enabling private, scalable messaging for Web3 by providing censorship resistance, adaptability, modular design, and a shared service network. 
```

Not sure what _metadata-resistant_ means. Do you mean _resistant to metadata harvesting_?

Waku is designed to enable communication between decentralized applications (dApps) in a peer-to-peer manner.  
It serves as an improvement and successor to Ethereum's Whisper protocol, offering better scalability and efficiency.

[Waku Relay](https://rfc.vac.dev/spec/11/), on the other hand, is the core component within the broader Waku framework.  
It is responsible for relaying messages between nodes in the Waku network, effectively serving as the message dissemination mechanism.  
While Waku encompasses a variety of functionalities and improvements for decentralized messaging, Waku Relay specifically focuses on the message propagation aspect within this larger system.

Finally, [Discv5 (Discovery v5)](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md) is a peer-to-peer networking protocol designed to facilitate node discovery in decentralized networks. 
It serves as an upgrade to earlier discovery protocols and is designed to be modular and extensible, allowing it to support various types of decentralized systems beyond just Ethereum.  
The protocol enables nodes to find each other, maintaining a distributed hash table.

> The scalability and performance of the Waku protocol are of critical importance. 

Reviewer: ```suggestion
The scalability and performance of the Waku protocol are of critical importance for any projects using Waku and intending to onboard thousands, if not millions, of users. 
```

- ##### Process-level monitoring
  Process-level monitoring offers detailed insights by tracking operational traits of processes within containers.
This method employs deep inspection of individual processes running inside a container by accessing */proc* kernel files to gather real-time process statistics.  
Reading from the */proc* filesystem offers a direct window into kernel statistics, providing comprehensive metrics on each process within the containers.
However, while it offers granular insights, process-level monitoring can be more resource-intensive and might not always capture overarching system behavior.   

### Performance Metrics

Each simulation lasted 3 hours to reach a steady state.  

- Hardware Level Metrics: Emphasis on memory usage, CPU consumption, disk I/O, and network I/O.

- Topology Level Metrics: Focuses on the efficiency of message propagation across the network, including metrics like message injection, propagation times, and message loss.

### Scalability  

To overcome scalability challenges, Wakurtosis employs a multi-node approach, running several nodes within one container.

> This method supports simulations with over 1,000 nodes on a single machine.  

Reviewer: The biggest network size mentioned below is 600 nodes. Why not using the full capacity of Wakurtosis?

For the two highest rates, bandwidth noticeably decreases from the smallest to moderate-sized networks. This counterintuitive drop suggests potential efficiencies or optimizable dynamics in intermediate networks.

However, in the larger 300- and 600-node configurations, the demands become more pronounced, especially from 300 to 600 nodes. While moderate networks may realize efficiencies, expansive networks exert greater transmission bandwidth demands.

The reception bandwidth (Rx) also manifests distinct patterns based on network size and message rate. For smaller networks, an intriguing trend emerges: across most rates (except 0 msg/s), bandwidth decreases consistently from 75 to 150 nodes. This indicates possible efficiencies or dynamics benefiting moderate configurations.

After this initial decrease, the bandwidth plateaus without significant growth even as rates rise. This contrasts sharply with the 600-node setup, where even without messaging, substantial bandwidth is consumed, comparable to active messaging in smaller networks.

> ![Total mean bandwidth usage without discovery mechanism](/static/img/wakurtosis_waku/non_discv5_bandwidth.png) 

Reviewer: I am not sure I understand the scale. The graph indicates _MB_ for the bandwidth

Is that the total bandwidth used over the 3 hours of message injection?

Would be good to provide some basis. For example, what is the baseline data volume we are injecting? 3 hours * 1 msg per second * msg size.

Also the graph has 5 lines but we stated 4 message rates (0.25, 0.5, 0.75, 1 messages per second).

Author: Yes, this is the average total bandwidth usage per node over the 3h hour simulation. I will add the pure payload lines into the plot to show how much data we are actually injecting into the network.

Author: @fryorcraken it's 5 series: 0.0, 0.25, 0.5, 0.75, and 1.0 msg/s. I just realised there is a missing legend. I ll fix that too.

<figcaption>

***Total mean bandwidth usage without discovery mechanism (baseline).***

</figcaption>

In addition to bandwidth, we examined average peak memory usage across network sizes and message loads. Memory use is fairly consistent without messaging, fluctuating between 20.3-20.5 MB. With messaging, only a slight increase is seen. Remarkably, at 1 msg/s, memory use remains modest, with the largest average of 21.95 MB in the 150-node setup. The difference between an idle and highly active network is under 1.5 MB.

![Mean peak memory usage without discovery mechanism](/static/img/wakurtosis_waku/non_discv5_memory.png) 

<figcaption>

***Mean peak memory usage without discovery mechanism (baseline).***

</figcaption>

#### Results with discovery mechanism (Discv5)

> With the discovery mechanism discv5, distinct resource usage patterns emerge. Transmission bandwidth remains relatively stable despite an eight-fold node increase, evidencing the protocol's transmission efficiency under discovery. However, reception bandwidth shows a different trend. From 75 to 300 nodes, a significant increase occurs, especially at 0 msg/s (95.83 to 254.7 Mb/s). The 300 to 600 node change is even more dramatic, with bandwidth surging to 497.47 Mb/s without messaging.

Reviewer: > with bandwidth surging to 497.47 Mb/s without messaging.

This seems worrying. Cc @jm-clius @alrevuelta

Reviewer 2: 497.47 Mb/s with discv5 and without messaging is indeed worring, but I have never seen this with https://github.com/waku-org/waku-simulator. How can we reproduce it?

Reviewer 3: > 497.47 Mb/s with discv5 and without messaging is indeed worring, but I have never seen this with https://github.com/waku-org/waku-simulator. How can we reproduce it?

If you want we can select one day to do a quick call and reproduce this. It was done with Wakurtosis. You can do it with 3~ commands. But the script to analyze the data is owned by @Daimakaimura. Also @Daimakaimura, are those 497.47 Mb/s constant over the entire simulation? Did you check when this happened? Because if it is a "constant" behaviour, we can test in in 5~ minutes run and will be easier to check.

Author: @AlbertoSoutullo those ~500 MB are the average total bandwidth usage per node over the whole simulation (i.e. 3h). @alrevuelta we were also surprised with that result so we re-run the simulation and got similar results, plot attached. We should indeed arrange a meeting and discuss about this because it is indeed weird. Also, keep in mind this also happens without DIscv5.
[analysis_hw.pdf](https://github.com/vacp2p/vac.dev/files/13067256/analysis_hw.pdf)
[analysis_hw.pdf](https://github.com/vacp2p/vac.dev/files/13067271/analysis_hw.pdf)

Additionally, employing a multi-node approach may result in losing node-level sampling granularity depending on the metrics infrastructure used, e.g. Cadvisor.

Nevertheless, Wakurtosis offers the flexibility to choose between this and a 1-to-1 simulation, catering to the specific needs of each test scenario. The results presented in this stidu are all 1-to-1 simulations &mdash; i.e., one node per container.

## Examining the Waku Protocol

### Simulation Setup

To evaluate Waku under varied conditions, we conducted simulations across a range of network sizes, topologies, and message rates. Each simulation lasted 3 hours to reach a steady state.

The network sizes explored included 75, 150, 300, and 600 nodes. 
We run simulations with discovery &mdash; Discv5 &mdash; and without discovery mechanisms &mdash; static network &mdash;. 

For Non-Discv5 simulations, we used static topologies with average node degrees of K=50. 
In simulations with Discv5, we set the max_peers parameter to 50 to approximate similar average degrees.

> To stress test message throughput, we simulated message rates of 1, and 10 messages per second. 

Author: Added "The message rates rates were always constant and independent of the network size." for clarity.

The reduced bandwidth at high message rates stems from simulation infrastructure limits rather than protocol inefficiencies. 
The no-load measurements provide insights into baseline protocol overhead costs that grow with network size. 

Transmission overhead appears to increase linearly, while reception overhead accelerates sharply beyond 300 nodes. 

![Total average bandwidth usage without discovery mechanism](/static/img/wakurtosis_waku/Baseline_Bandwidth.png) 

<figcaption>

***Total average bandwidth usage without discovery mechanism (baseline).***

</figcaption>

In addition to bandwidth, we also examined average peak memory usage under the different network sizes and messaging loads. 

![Average peak memory usage without discovery mechanism](/static/img/wakurtosis_waku/Baseline_Memory.png) 

<figcaption>

***Average peak memory usage without discovery mechanism (baseline).***

</figcaption>

With no load, memory usage remained consistent around 20-21 MB across all network scales. 
Under 1 message per second, average memory usage increased slightly to 21-22 MB. 

> The highest memory usage of 23-25 MB was seen with 10 messages per second, especially at 150 nodes. 

Author: No, I do not have an explanation for this memory peak. Also if you check the scale in the y-aixs (+/- 4MB) you'll see that despite it shows a peak it actually not that big if you zoom out slightly. But looks to me like sampling noise.

This aligns with expectations that higher messaging loads require more memory for message processing and routing. 
However, the differences are relatively small, with only around a 20% increase from no load to 10 msgs/s. 
This suggests the protocol has a fairly fixed memory overhead cost, with incremental increases as more messages are handled. 
Overall, the memory usage appears stable and scales reasonably well, without any dramatic growth as network size expands. 
This consistency indicates efficient memory management and low overhead from a protocol perspective.

The baseline simulations displayed efficient baseline overhead and scaling for memory, transmission, and reception as message rates grew or network size expanded.

#### With discovery mechanism (Discv5)

With the discovery mechanism, we again see interesting trends in resource usage across different network sizes and traffic loads. 

> Transmission bandwidth remainded fairly stable with more nodes.

Reviewer 4: What does this mean? Does this tell us something about Waku messaging protocol, our chosen discovery technique and network size?

## Introduction
The scalability and performance of the Waku protocol are subjects of critical importance to the VAC-DST team.
To explore these facets with high granulatiry across a wide range of scenarios we turned to Wakurtosis, a bespoke simulation framework developed internally. 
By studying various network sizes, message rates, and peer discovery setups we aimed to better understand the protocol's capabilities and limitations, hence aspects that could benefit from further optimisation.

## Understanding Wakurtosis
Wakurtosis is a robust simulation framework which integrates Docker and Kurtosis to create a simulation environment which allows to run highly granular, large scale simulations with a variety of traffic and network patterns. 

> At the core of Wakurtosis is Kurtosis — an orchestration tool responsible for managing containers, known as services, within isolated environments called enclaves. 

Author: Added relevant links in throughout the document

- Hardware Level Metrics: Emphasis on memory usage, CPU consumption, Disk I/O, and Network I/O.
- Topology Level Metrics: Focuses on the efficiency of message propagation across the network, including metrics like message injection, propagation times, and message loss. 

### Scalability
To overcome scalability challenges, Wakurtosis employs a multi-node approach, running several nodes within one container. 
This method supports simulations with over 1,000 nodes on a single machine. 
However, this can introduce unforeseen network effects, potentially affecting some metrics.
For instance, running several nodes per container can alter propagation times, as nodes grouped within the same container may exhibit different messaging behavior compared to a true node-to-node topology
Additionally, employing a multi-node approach may result in losing node-level sampling granularity depending on the metrics infrastructure used, e.g. Cadvisor.
Nevertheless, Wakurtosis offers the flexibility to choose between this and a 1-to-1 simulation, catering to the specific needs of each test scenario. 
The results presented in here are 1-to-1 simulations.

## Examining the Waku Protocol

> 

Author: Added to the introduction with links.

### Simulation Setup
To evaluate Waku under varied conditions, we conducted simulations across a range of network sizes, topologies, and message rates. 
Each simulation lasted 3 hours to rreach a steady state. 
The network sizes explored included *75*, *150*, *300*, and *600* nodes. 
For non-Discv5 simulations, we used static topologies with average node degrees of *K=3*, *K=13*, and *K=50*. 
In simulations with Discv5, we set the *max_peers* parameter to *12* and *50* to approximate similar average degrees. 
To stress test message throughput, we simulated message rates of *1*, *10*, and *100* messages per second. 
This combination of network sizes, topologies, message rates, and hardware configurations enabled us to comprehensively evaluate Waku's performance and scalability boundaries under diverse conditions.

> 

Author: Added the following to clarify this point:

> It's important to note that the Non-Discv5 scenarios presented in this study serve as a theoretical baseline for comparison and are not meant to represent real-world conditions. In a live environment, some form of peer discovery mechanism would be necessary for the functioning of the network. Mechanisms like 'rendezvous' would also introduce additional bandwidth costs. Therefore, the intention behind including Non-Discv5 simulations is not to compare their performance directly with Discv5 but rather to establish a fundamental baseline against which the added complexities and scalling costs of employing a discovery mechanism like Discv5 can be better understood.

### Results
To provide a clear visualization of bandwidth usage in all the cases we studied, we've presented the results as multipliers over the pure payload for both reception (Rx) and transmission (Tx) in two heatmaps:

![Bandwidth effiencies multipliers over pure payloads in transmission](/static/img/wakurtosis_waku/waku_bandwidth_rx_efficiency.png)
![Bandwidth efficiencies multipliers over pure payloads in reception](/static/img/wakurtosis_waku/waku_bandwidth_tx_efficiency.png)

Despite certain scalability challenges, Wakurtosis has proved effective in simulating a Waku network with over *600* nodes on a single machine. 
Our analysis under baseline conditions --- static network with average degrees of *K=3*, *K=13*, and *K=50* --- revealed good scalability and stability from a memory standpoint. 
However, there was evidence of increased memory usage with higher message rates, indicating potential areas for optimization. 
Bandwidth utilization exhibited minor fluctuations as the network size increased, and efficiency in bandwidth usage improved with an increase in the message rate. 

> Some inefficiencies were noted in larger networks, suggesting room for improvement but overall Waku displayed very stable behaviour in terms of memory and bandwidth usage.

Author: Added some number and general trends regarding memory usage for both baseline runs and discv5 runs.

#### Impact of Discovery v5
To shed light on the effects of Discovery v5, we focus on the denser configurations for both non-Discv5 and Discv5 with *K=50*.
The addition of Discv5 leads to higher memory consumption, especially as the network size increases.
This is likely due to the additional routing information that needs to be stored and maintained.
When examining the effects on bandwidth utilization, we see varying results.
For reception rates, with lower message rates we observe that Discv5 consistently utilizes more bandwidth than the corresponding non-Discv5 case.
However, those differences become less pronounced with higher message rates.

> For transmission rates, both cases show similar performance and both exhibit huge improvements in transmission efficiency at higher message loads.

Reviewer 4: One more note, to reiterate: we wouldn't expect to see an efficiency that's higher than around 1/6. Where this is the case, we should assume there might be losses due to infrastructure (or other) limits.

## Conclusion
This study underscores the Waku protocol’s resilience and scalability across varied conditions, but also highlights the limitations of Wakurtosis and the need for a more robust simulation infrastructure for demanding scenarios. 
The protocol’s robustness, evidenced by the absence of message loss, notable stability across network sizes and traffic loads, is a notable takeaway. 

> The addition of Discv5 generally leads to higher memory and bandwidth usage throughout the majority of scenarios. 

Author: Slightly tuned the paragraph and also added the discv5 / non-discv-v5 clarification in the simulation setup section

---
layout: post
name: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'
title: 'Scaling the Waku Protocol: A Performance Analysis with Wakurtosis'
date: 2023-08-08 12:00:00
authors: Daimakaimura
published: true
slug: wakurtosis-waku-scallability-simulations
categories: waku, wakurtosis
---

# Scaling the Waku Protocol: A Performance Analysis with Wakurtosis

<!--truncate-->

## Introduction
The scalability and performance of the Waku protocol are subjects of critical importance to our team. 
To explore these facets in depth --- i.e. high granularity --- we turned to Wakurtosis, a bespoke simulation framework developed by the VAC-DST team. 

> By studying various network sizes, message rates, and peer discovery setups we aimed to better understand the protocol's capabilities and limitations, hence aspects that could benefit from further.

Reviewer: ```suggestion
By studying various network sizes, message rates, and peer discovery setups we aimed to better understand the protocol's capabilities and limitations, hence aspects that could benefit from further <missing word>.
```

At the core of Wakurtosis is Kurtosis — an orchestration tool responsible for managing containers, known as services, within isolated environments called enclaves. 
These enclaves house virtual networks and their respective containers. In addition to this, several external modules developed in-house expand some of Kurtosis's limited functionalities:
- Network Generation Module (Gennet): Initiates and configures networks for the simulation. It's highly modular, supporting the integration of multiple protocols and node traits.
- Packet Injection Module (WLS): Allows for the insertion of custom data packets, thereby enabling varied traffic patterns and stress tests on the simulations.
- Analysis Module: Captures and provides insights into resource usage, network behaviors and protocol interactions throught the enclave.

### Data Collection
Wakurtosis ensures the accuracy of its data by leveraging multiple sources for hardware metrics:

##### Cadvisor (a Google tool)

> Cadvisor providess detailed metrics on resource usage and performance characteristics of Docker containers. 

Reviewer 4: ```suggestion
Cadvisor provides detailed metrics on resource usage and performance characteristics of Docker containers. 
```

This method employs deep inspection of individual processes running inside a container by accessing */proc* kernel files to gather real-time process statistics. 
Reading from the */proc* filesystem offers a direct window into kernel statistics, providing comprehensive metrics on each process within the containers. 
However, while it offers granular insights, process-level monitoring can be more resource-intensive and might not always capture overarching system behavior.

### Performance Metrics
Data sources like Cadvisor/Prometheus, Docker Stats, and process-level monitoring enhance the reliability of Wakurtosis, offering a comprehensive view of system performance.

- Hardware Level Metrics: Emphasis on memory usage, CPU consumption, Disk I/O, and Network I/O.
- Topology Level Metrics: Focuses on the efficiency of message propagation across the network, including metrics like message injection, propagation times, and message loss. 

### Scalability
To overcome scalability challenges, Wakurtosis employs a multi-node approach, running several nodes within one container. 
This method supports simulations with over 1,000 nodes on a single machine. 

> However, this can introduce unforeseen network effects, potentially affecting some metrics. 

Reviewer 4: Is there a one sentence example of one of these effects?

For non-Discv5 simulations, we used static topologies with average node degrees of *K=3*, *K=13*, and *K=50*. 
For Discv5 simulations, we configured max_peers settings of 12 and 50 to approximate similar average degrees. 
To stress test message throughput, we simulated message rates of 1, 10, and 100 messages per second. 
This allowed us to observe performance under low, moderate, and high traffic loads. We leveraged Wakurtosis' ability to run multi-node simulations to achieve networks at the scale of 600 nodes. 
This combination of network sizes, topologies, message rates, and hardware configurations enabled us to comprehensively evaluate Waku's performance and scalability boundaries under diverse conditions.

### Results
Despite certain scalability challenges, Wakurtosis has proved effective in simulating Waku with over 1,000 nodes on a single machine. 
We examined the behaviours of the Waku protocol with multiple network sizes, message rates and peer discovery strategies. 
Our analysis under baseline conditions --- static network with average degrees of *K=3* and *K=13* --- revealed promising scalability and stability from a memory standpoint. 

> Still, there was evidence of increased memory usage with higher message rates, indicating potential areas for optimization. Bandwidth utilization exhibited minor fluctuations as the network size increased, and efficiency in bandwidth usage improved with an increase in the message rate. 

Reviewer 4: "bandwidth usage improved with an increase..."

I think this is simply an effect of hardware constraints, so perhaps need to be clarified or left out. Just in case people believe we're capable of miracles. :)

Author: LOL ... indeed!

> Some inefficiencies were noted in larger networks, suggesting room for improvement but overall Waku displayed very stable behaviour in terms of memoery and bandwidth usage.

Reviewer 4: ```suggestion
Some inefficiencies were noted in larger networks, suggesting room for improvement but overall Waku displayed very stable behaviour in terms of memory and bandwidth usage.
```

> ![Bandidth effiencies multipliers over pure payloads in transmission](/static/img/wakurtosis_waku/waku_bandwidth_rx_efficiency.png)

Reviewer: ```suggestion
![Bandwidth efficiencies multipliers over pure payloads in transmission](/static/img/wakurtosis_waku/waku_bandwidth_rx_efficiency.png)
```

![Bandwidth efficiencies multipliers over pure payloads in reception](/static/img/wakurtosis_waku/waku_bandwidth_tx_efficiency.png)

> #### Impact of Discovery v5

Reviewer: Looks like this whole section is commented out.

> <!--When incorporating Discovery v5 into the Waku protocol, we initially observed marked differences in performance metrics, particularly concerning memory consumption and bandwidth usage. However, after adjusting our non-Discv5 simulation parameters to better mirror the natural conditions of Discv5 (namely, by focusing on an average degree closer to 15), these disparities became less pronounced.-->

Reviewer 4: The framing of discv5 here (and previously) can perhaps be a bit clearer. It's not really "integrated" into the protocol or some inherent feature - it's simply another protocol running alongside Waku that helps a Waku node discover and make connections to other peers in the network.

Author: Apologies. I fixed it to reflect the fact that it is a separate protocol, ie an addition to the setup.

<!--truncate-->

## Introduction

> The scalability and performance of the Waku protocol are subjects of critical importance to our team. To explore these facets in depth --- i.e. high granularity --- we turned to Wakurtosis, a bespoke simulation framework developed by the VAC-DST team. By studying various network sizes, message rates, and peer discovery setups we aimed to better understand the protocol's capabilities and limitations, hence aspects that could benefit from further.

Reviewer: you have "our team" and "VAC-DST team" in the same paragraph. IMO it sounds weird as I expected "our team" and "VAC-DST" to be the same. Or Am I mistaken?
